src/main/scala/mesosphere/marathon/core/storage/repository/impl/CachedTemplateRepository.scala

```diff
@@ -0,0 +1,91 @@
+  *
+  * All template versions are saved in the cache using [[CachedTemplateRepository#toCacheKey()]]
+  * method e.g. a template with pathId=/foo/bar and version=1 will be saved using the key `foo_bar:1`.
+  *
+  * Basically this class wraps [[TemplateRepository]] methods with the calls to cache. [[read()]] method goes directly to cache
+  * and the element will be loaded if not already exists.
+  *
+  * Note that this implementation is tread-safe and provides serializability guarantees e.g. two independent [[create()]]
+  * calls are not guaranteed to be executed in any particular order the same way one should not expect two independent futures
+  * to have an order.
+  *
+  * Known pitfalls: an operation e.g. [[delete()]] may fail leaving the cache and the underlying store in an inconsistent
+  * state: e.g. the template was removed from the store but not from the cache. This is very unlikely but still possible.
+  * In this case the operation's future will fail and it is the responsibility of this repository user to react accordingly.
+  * [[delete()]] operation are idempotent and can be safely repeated while [[create()]] operation will increment the template
+  * version and create a new one. In this case the "failed" template version will have to be removed.
+  *
+  * @param store underlying instance of [[ZooKeeperPersistenceStore]]
+  * @param cacheMaximumSize maximum cache size
+  * @param cacheExpireAfter cached element expiration duration
+  */
+
+@SuppressWarnings(Array("all")) // async/await
+//format:off
+class CachedTemplateRepository(
+    override val store: ZooKeeperPersistenceStore,
+    val cacheMaximumSize: Int = 10000,
+    val cacheExpireAfter: Duration = 30.minutes)(implicit ec: ExecutionContext, mat: ActorMaterializer)
+  extends TemplateRepository(store) {
+
+  import CachedTemplateRepository._
+
+  // We can cache aggressively since template versions are immutable and once created/fetched can
+  // stay in cache forever. The key is template's relative storage path: {PathId.safePath}:{Version} e.g. `eng_sleep:1`
+  val cache: AsyncLoadingCache[String, Template] =
+    Scaffeine()
+      .recordStats()
+      .expireAfterWrite(cacheExpireAfter)
+      .maximumSize(cacheMaximumSize.toLong)
+      .buildAsyncFuture { (key: String) =>
+        val (pathId, version) = fromCacheKey(key)
+        super.read(pathId, version)
+      }
+
+  override def create(spec: Spec): Future[Template] = async {
+    val versioned = await(super.create(spec))
+    cache.put(toCacheKey(versioned), Future.successful(versioned))
+    versioned
+  }
+
+  override def read(pathId: PathId, version: Int): Future[Template] =
+    cache.get(toCacheKey(pathId, version))
+
+  override def delete(pathId: PathId): Future[Done] = async {
+    val versions = await(super.versions(pathId)) // Get all existing template versions
+    val deleted = await(super.delete(pathId)) // Delete template from the store
```

Comment:
How is it ensured that there is no race condition in situations like this: getting `versions`, and then deleting the template, when a new version is created in the meantime?

Reply:
I'm glad someone is paying attention! I was thinking about this scenario and tried to synchronize this through `scaffeine` cache, but unfortenately it seems to be  impossible due to _scaffeine_ implementation details (`AsyncLoadingCache` doesn't work together with deleting/updating elements through cache).

To be precise about the problem at hand: assuming we have 2 independent operations: `create(/foo)` and `delete(/foo)` there are two trivial cases where either `a) create() -> delete()` or `b) delete() -> create()` happen after previous operation completed successfully. Here we either end up with `a)` everything deleted or `b)` a new version of template is created (with version counter started with `1` again). Both cases are completely legal.

Now to not-trivial cases where both operations interleave:
- `delete()` has 3 major steps:
  - (1) <-
  - get all existing `versions`
  - (2) <-
  - `delete` parent `/foo` node (and all templates versions in it)
  - (3) <-
  - `invalidateAll` versions from the cache

A concurrent `create(/foo)` call can happen in `(1), (2), (3)` places. Let's see what will happen then:
1. If `create` call happens here, the resulting version will be simply deleted with everything else in the subsequent steps *in the ZK*. However since `create` is also 2-steps operation (store in ZK and store in cache) we might end up with a cached version even after `delete` has finished.
2. Same logic applies so that we might end up with a version in cache but not in ZK. 
3. Here we already deleted `/foo` node from the ZK, so a new template version will be stored successfully. However the subsequent caching might be invalidated, so we'll end up with a version in ZK but not in cache. This is however fine, since cache will load it on the next _read_ request.

So, as we can see, in cases `(1)` and `(2)` we might end up with a version in cache which does not exist in the store (which is clearly bad). However I haven't found a lock-free solution to the problem (synchronizing everything is an "easy fix").

Reply:
Yeah, I don't think there is an "easy fix".

This is one reason I proposed the in-memory state authority which had a primary job to enforce foreign key constraints. It's effectively synchronizing everything, but, performing each serial operation in memory. Multi-entity deletes like this could be done in a single frame, and subsequent modifications could be queued up in a zookeeper transaction.

Reply:
An optimistic lock strategy may be possible? Just spitballing here:

1) Query the current template version as stored
2) Query the versions
3) Delete / invalidate / etc
4) Query the current template version again. If it changed from step 1, then repeat steps

Reply:
> An optimistic lock strategy may be possible?

Well, yeah, it might, but I'm not looking for complete serializability of operations. E.g. in the above example, `delete(/foo)` will always try to clean everything before it finishes even if a concurrent `create(/foo)` operation exists. What I'm aiming for instead, is for both operations to be atomic where `a) create() -> delete()` or `b) delete() -> create()` can have any order (should be enforced by the caller and not by the repo). 
The important part is that both operations don't interleave and/or leave a consistent cache/store behind. `synchronize` seems to be the easiest fix, but I was hoping for a more elegant idea.

Reply:
Just thinking out loud here - we will not have this problem if deleting a node in ZK would return a list of deleted children (or `[]` otherwise). Then the method above could be:
```scala
  override def delete(pathId: PathId): Future[Done] = async {
    val versions = await(super.delete(pathId)) // Delete template from the store and return versions
    cache.synchronous().invalidateAll(versions.map(toCacheKey(pathId, _))) // Invalidate all cached versions
    Done
  }
```
Here we would always only invalidate versions that were previously deleted, but ZK doesn't have this functionality... 🤔

Reply:
What about plain old `syncronized`?

Reply:
Alternatively, we could delete all versions *in one transaction*:
```scala
  override def delete(pathId: PathId): Future[Done] = async {
    val versions = await(super.versions(pathId)) // Get all existing template versions
    val deleted = await(super.transaction(versions.map(super.deleteOp(_)) // A bit of pseudocode: delete all verasions from the store in a transaction
    cache.synchronous().invalidateAll(versions.map(toCacheKey(pathId, _))) // Invalidate all cached versions
    Done
  }
```
but this would leave a dangling parent node in the store.